docs: Add config tutorial (#4231)
nf-core/website: sites/docs/src/content/docs/running/tutorials/configure-a-pipeline-run.md

---
title: Configure a pipeline run
subtitle: A hands-on tutorial for configuring nf-core/demo with environment variables, parameters, and config files
shortTitle: Configure a pipeline run
---

nf-core pipelines are shaped by multiple levels of of configuration: environment variables that tune Nextflow itself, parameters that tune the pipeline, and config files that tune execution.
Using [`nf-core/demo`](https://nf-co.re/demo), this tutorial explains each layer as worked examples.

![nf-core/demo subway map](../../../../assets/images/get_started/nf-core-demo-subway.png)

By the end you will have:

> - Used `NXF_*` environment variables to control Nextflow's runtime.

Reviewer: Yeah I would move this later, it's more advanced IMO - probably after custom configs normally.

Author: Moved.

- Supplied pipeline parameters on the command line and from a `params.yaml` file.
- Activated profiles to switch between container engines and test scenarios.
- Layered a custom `custom.config` file with process-specific overrides.
- Used `ext.args` (the convention for `conf/modules.config`) to extend a tool's command line without editing the module.
- Understood which configuration source overrides which.

:::note{title="Prerequisites"}
You will need the following to get started:

- An internet connection

> - [Nextflow version 24.04.0 or later](../get_started/environment_setup/nextflow.md)

Reviewer: ```suggestion
- [Nextflow version 25.10.4 or later](../get_started/environment_setup/nextflow.md)
```

This is the current minimum version in the template, if we think this is important

Author: Agreed. Updated.

- A container engine, such as Docker, Singularity, or Conda. See [Software dependencies](../../get_started/environment_setup/software-dependencies.md)

:::

## Baseline run

Before changing anything, run the pipeline with no custom configuration so you have something to compare against.
This baseline run relies entirely on `nf-core/demo`'s built-in defaults.
No environment variables, custom parameters, no extra config files are set.
In each of the following steps you will replace one of these defaults and observe the effect.

Start with a plain test invocation as your reference point:

```bash
nextflow run nf-core/demo -profile test,docker --outdir results
```

Replace `docker` with your preferred software dependency manager (for example, `singularity` or `conda`).

> This run uses three defaults that you will override in the next steps:

Reviewer: ```suggestion
This run is using three defaults that you will override in the next steps:
```

Or you could say 'In the background, this run uses three`, to make it clearer this is the default behaviour of Nextflow, that the user did not actively select some of these.

Author: Added.

> - The pipeline's bundled [`nextflow.config`](https://github.com/nf-core/demo/blob/master/nextflow.config) and the [`test`](https://github.com/nf-core/demo/blob/master/conf/test.config) and `docker` profiles

Reviewer: ```suggestion
- The pipeline's internal [`nextflow.config`](https://github.com/nf-core/demo/blob/master/nextflow.config) and the [`test`](https://github.com/nf-core/demo/blob/master/conf/test.config) and `docker` profiles
```

I feel `bundled` might be too much of an english colloquiualism

> - Default pipeline parameters (input samplesheet, MultiQC title, and so on). See the [nf-core/demo parameters page](https://nf-co.re/demo/parameters) and [`nextflow_schema.json`](https://github.com/nf-core/demo/blob/master/nextflow_schema.json)

Reviewer: ```suggestion
- Default pipeline parameters (path to input samplesheet, the customisable MultiQC title and so on). See the [nf-core/demo parameters page](https://nf-co.re/demo/parameters) and [`nextflow_schema.json`](https://github.com/nf-core/demo/blob/master/nextflow_schema.json)
```

- Your default Nextflow runtime settings (e.g., work directory and Nextflow version)

:::tip

> Keep the `results/` directory from this run. You can compare it to the output of later steps to confirm your configuration changes took effect.

Reviewer: ```suggestion
Keep the `results/` directory from this run.
You can compare it to the output of later steps to confirm your configuration changes took effect.
```

:::

> ## Configure with environment variables

Reviewer: Initial thought: these are things people rarely touch (in my experience) so being the first thing and something they will not commonly use might be a bit jarring - I would suggest putting this a bit further down.

`NXF_*` environment variables are the outermost configuration layer.
Nextflow reads them from your shell as it starts, before any config file or parameter is parsed, so they're the right place for settings that don't change between runs.
For example, which Nextflow version to use, where the work directory and container cache live, and whether to operate offline.
They configure **Nextflow itself**, not the pipeline.
You cannot set pipeline inputs or outputs this way.

You set them like any other shell variable: with `export`, inline before a single command, or by adding them to your shell's startup file.

1. Pin a Nextflow version and redirect the work directory for this run:

   ```bash

>    export NXF_VER=24.10.5

Reviewer: Hrmm, his doesn't work for me, I get:

```
Failed requirement - Plugin nf-schema@2.6.1 requires Nextflow version >=25.10.0 (current 24.10.5)
```

Maybe we should bump versions throughout?

   export NXF_WORK=$HOME/nf-demo-work

>    nextflow run nf-core/demo -profile test,docker --outdir results

Reviewer: ```suggestion
   nextflow run nf-core/demo -profile test,docker --outdir results_customvariable
```

   ```

   The version line at the top of stdout (and in `.nextflow.log`) should now report `24.10.5`, and intermediate files should appear under `$HOME/nf-demo-work/` instead of `./work/`.

2. To set a variable for a single command without exporting it, prefix the invocation:

   ```bash
   NXF_VER=24.10.5 nextflow run nf-core/demo -profile test,docker --outdir results
   ```

> 3. To persist a variable across sessions, add the export to your shell config:

Reviewer: I think this should go into a tip box, or removed entirely,  I don't think this should be a recommended pattern because it can cause problems later if someone forgets they set it (speaking from experience)

   ```bash title=".bashrc"
   export NXF_VER=24.10.5
   ```

:::tip
The `NXF_*` variables nf-core users reach for most often are:

- `NXF_VER`: Pin a specific Nextflow version

> - `NXF_HOME`: Override Nextflow's home directory (defaults to `$HOME/.nextflow`)

Reviewer: Would be good to have a link to a page on the NXF docs that says what goes in there, many users may not ever know its there or what it does... I tried looking quickly now but can't find such a description on https://docs.seqera.io... the closest I can find what is inside NXF_HOME is https://docs.seqera.io/nextflow/cli#remote-pipelines

- `NXF_WORK`: Override the work directory for intermediate files

> - `NXF_OFFLINE`: Run without contacting the network (see [Running pipelines offline](../run-pipelines-offline.md))

Reviewer: ```suggestion
- `NXF_OFFLINE`: Run without contacting the internet (see [Running pipelines offline](../run-pipelines-offline.md))
```

> - `NXF_SINGULARITY_CACHEDIR` / `NXF_APPTAINER_CACHEDIR`: Reuse downloaded container images across runs

Reviewer: ```suggestion
- `NXF_SINGULARITY_CACHEDIR` / `NXF_APPTAINER_CACHEDIR`: Container store location to allow reuse across runs
```

> For the full list, see the [Nextflow environment variables reference](https://docs.seqera.io/nextflow/reference/env-vars).

Reviewer: Ok you have this, but I feel for NXF_HOME this is not sufficiently descriptive of what its used for

:::

:::note

> `NXF_*` variables are distinct from pipeline parameters. They configure how Nextflow runs, not what the pipeline does. You cannot set `--input` or `--outdir` with an environment variable.

Reviewer: ```suggestion
`NXF_*` variables are distinct from pipeline parameters.
They configure how Nextflow runs, not what the pipeline does.
You cannot set `--input` or `--outdir` with an environment variable.
```
You had a sentence at the top of this section with this, given this is a tutorial which generally can be slightly more informal , maybe it makes sense to say 'Remember:' or something

:::

## Configure with parameters

> Pipeline parameters are the knobs the pipeline itself exposes.

Reviewer: ```suggestion
Pipeline parameters are the knobs and switches the pipeline offers to users to control how the different steps of the pipeline executes.
```

'Knob' again felt slightly too colloquiual, and `exposes` can mean different things (and also doesn't say to whom)

For example, inputs, outputs, skip flags, reference data choices, and tool toggles.
They're documented per pipeline.
For `nf-core/demo`, see the [parameters reference](https://nf-co.re/demo/parameters) and [`nextflow_schema.json`](https://github.com/nf-core/demo/blob/master/nextflow_schema.json).

You can supply them two ways:

- Directly on the command line
- From a `params.yaml` or `params.json` file.

The two forms are interchangeable, and you can mix them.

:::important
CLI flags override values in a parameter file.
:::

1. Override a single parameter on the command line:

   ```bash

>    nextflow run nf-core/demo -profile test,docker --outdir my_results

Reviewer: ```suggestion
   nextflow run nf-core/demo -profile test,docker --outdir results_customcommandline
```

   ```

   Output now goes to `my_results/` instead of `results/`.

2. For longer parameter sets, create a `params.yaml` file:

   ```yaml title="params.yaml"

>    outdir: my_results

Reviewer: ```suggestion
   outdir: results_customparamsfile
```

>    multiqc_title: "nf-core/demo configured run"

Reviewer: ```suggestion
   multiqc_title: "nf-core/demo parameter file configured run"
```

   ```

3. Apply it with `-params-file`:

   ```bash
   nextflow run nf-core/demo -profile test,docker -params-file params.yaml
   ```

   Open `my_results/multiqc/multiqc_report.html` to confirm the report title was set.

:::tip

> Both JSON and YAML are supported for parameter files.

Reviewer: ```suggestion
Both JSON and YAML formats are supported for parameter files.
```

Reviewer: I would link to the NXF docs here too, but I can only find one with a JSON example: https://docs.seqera.io/nextflow/cli#pipeline-parameters

Parameter files make runs easier to reproduce and share.
You can publish them alongside results so others can re-run the exact same configuration.
:::

:::warning
nf-core pipeline parameters must be supplied on the command line or via `-params-file`.
Setting a parameter inside a custom `.config` file (for example, `params { outdir = '...' }`) will **not** override the pipeline's defaults.

> Use config files for executor and resource settings, and parameter files for pipeline parameters.

Reviewer: ```suggestion
Use config files for executor and computational resource settings, and parameter files for pipeline parameters.
```

:::

## Configure with config files

Config files are the most flexible configuration layer.
They control both **where** the pipeline runs and **how** it runs.

> For example, which executor submits jobs (e.g., `local`, `slurm`, `awsbatch`), how much CPU and memory each process should request, what Nextflow does when a process fails, which container registry to pull from, where intermediate work lives, how to apply different settings to specific steps, and much more.

Reviewer: ```suggestion
For example, which executor submits jobs (e.g., `local`, `slurm`, `awsbatch`), how much CPU and memory each process should request, what Nextflow does when a process fails, which container registry to pull from, where intermediate working files lives, how to apply different settings to specific steps, and much more.
```

> They're the right place for any setting that depends on your infrastructure rather than on what the pipeline does scientifically.

Reviewer: Nice way of saying it!

Nextflow assembles its final configuration from several config files.
Nextflow picks up some config files automatically, some you switch on by name with `-profile`, and others you point it at explicitly with `-c`.

The next four sections cover all three approaches and then show how to target specific processes inside any config file.

### Auto-loaded `nextflow.config` files

Nextflow looks for config files in three places without being told.
From lowest to highest priority, they are:

> 1. **The pipeline's own `nextflow.config` in the project directory.** Bundled with every nf-core pipeline and downloaded with the rest of the pipeline code. For example, [`nf-core/demo`'s `nextflow.config`](https://github.com/nf-core/demo/blob/master/nextflow.config), which loads [`conf/base.config`](https://github.com/nf-core/demo/blob/master/conf/base.config) and declares every shipped profile. **Don't edit this file**. See the warning in [Configuration options](../configuration/configuration-options.md).

Reviewer: ```suggestion
1. **The pipeline's own `nextflow.config` in the project directory.** Included with every nf-core pipeline's source code. For example, [`nf-core/demo`'s `nextflow.config`](https://github.com/nf-core/demo/blob/master/nextflow.config), which loads [`conf/base.config`](https://github.com/nf-core/demo/blob/master/conf/base.config) and declares every shipped profile. **Don't edit this file**. See the warning in [Configuration options](../configuration/configuration-options.md).
```

2. **`$HOME/.nextflow/config`.** Your personal config, applied to every Nextflow run you launch. Good for settings that should follow you across all pipelines, such as your Singularity cache location or a default executor for your workstation.
3. **`nextflow.config` in the launch directory.** Applied automatically whenever you run Nextflow from that directory. Good for per-project or per-working-directory settings you don't want to retype each time.

Add a `nextflow.config` in your current directory:

```groovy title="nextflow.config"
process {
  cpus = 2

>   memory = 4.GB

Reviewer: ```suggestion
  cpus = 1
  memory = 2.GB
```

Reviewer: Although whatever I set these values to it doesn't seem to change the stuff in the piplein_info report 🤔 but maybe somethign to check

}
```

Re-run the baseline command. No `-c` flag is needed, Nextflow picks the file up automatically:

```bash

> nextflow run nf-core/demo -profile test,docker --outdir results

Reviewer: ```suggestion
nextflow run nf-core/demo -profile test,docker --outdir results_customnextflowconfig
```

```

> 

Reviewer: We dont' have a away to show a user if it worked or not - maybe look inside `results/pipeline_info/execution_report`?

:::tip
Use `$HOME/.nextflow/config` for personal defaults that follow you across machines, and a launch-directory `nextflow.config` for project-specific settings.
Reserve `-c` (covered below) for one-off overrides and shared team configs that should live alongside your pipeline command.
:::

### Activate bundled settings with profiles

> A **profile** is a named bundle of configuration that lives inside a config file under the `profiles` scope.

Reviewer: ```suggestion
A **profile** is a named bundle of configuration that lives inside a config file under the `profiles` scope.
They allow you to switch between different sets of configurations stored in a single config file.
```

You activate one or more with the `-profile` flag.

> You've been using profiles since [Baseline run](#baseline-run). `-profile test,docker` activates two: `test` (a small public dataset) and `docker` (use Docker to manage software).

Reviewer: ```suggestion
You've actually been using profiles since [Baseline run](#baseline-run).
`-profile test,docker` activates two: `test` (a small public dataset) and `docker` (use Docker to manage software).
```

I would prefer a `!` at the end of the sentence, but I know it's against teh style guide

> Every nf-core pipeline ships with a standard set of profiles:

Reviewer: ```suggestion
Every nf-core pipeline comes with a standard set of profiles:
```

- **Software profiles**: `docker`, `singularity`, `apptainer`, `podman`, `conda`, `charliecloud`, `shifter` (one per container engine or environment manager).

> - **Test profiles**: [`test`](https://github.com/nf-core/demo/blob/master/conf/test.config) (a small dataset for quick verification) and [`test_full`](https://github.com/nf-core/demo/blob/master/conf/test_full.config) (the full-size dataset used in CI).

Reviewer: ```suggestion
- **Test profiles**: at a minimum [`test`](https://github.com/nf-core/demo/blob/master/conf/test.config) (a small dataset for quick verification) and [`test_full`](https://github.com/nf-core/demo/blob/master/conf/test_full.config) (the full-size dataset used in CI).
```

Reviewer: I'll need to continue another day from here

  :::note

>   The [`test_full`](https://github.com/nf-core/demo/blob/master/conf/test_full.config) is also a very small test dataset that can be used for testing.

Reviewer: ```suggestion
  The [`test_full`](https://github.com/nf-core/demo/blob/master/conf/test_full.config) profile in nf-core/demo is also a very small test dataset that can be used for testing.
  In other pipelines these can be much larger, but produce realistic output.
```

  :::

> - **Institutional profiles**: contributed to [nf-core/configs](https://github.com/nf-core/configs) and loaded automatically by every nf-core pipeline. Activate one with `-profile <institution>` if your cluster has one.

Reviewer: ```suggestion
- **Institutional profiles**: contributed to [nf-core/configs](https://github.com/nf-core/configs) and loaded automatically by every nf-core pipeline. Activate one with `-profile <institution>` if your cluster has one. See the nf-core website for a list of [available institutional configs](https://nf-co.re/configs).
```

Reviewer: Can drop that suggestion if you want to reserve the information for the next section

Reviewer: ```suggestion
- **Institutional profiles**: contributed to [nf-core/configs](https://github.com/nf-core/configs) and loaded automatically by every nf-core pipeline. Activate one with `-profile <institution>` if your cluster has one - see [Use shared institutional configs](#use-shared-institutional-configs).
```

Instead maybe?

> Combine profiles with commas, and remember that order matters — later profiles override earlier ones where they overlap:

Reviewer: ```suggestion
Combine profiles with commas, and remember that order matters — later profiles override earlier ones where they overlap.
In this example, options in the `docker` profile will override any of the options with the same name as in `test`. 
```

Reviewer: Not sure 'name' is the best word to use here, but I couldn't find a good alternative and not repeat 'options'.

```bash
nextflow run nf-core/demo -profile test,docker --outdir results
```

You can also define your own profile for settings you'd like to reuse.

> Add a `myserver` profile to your launch-directory `nextflow.config`:

Reviewer: ```suggestion
Append a `myserver` profile to your existing launch-directory `nextflow.config`:
```

```groovy title="nextflow.config"
profiles {

>   myserver {

Reviewer: ```suggestion
  mymachine {
```

Reviewer: To make it more generic to where people are running, and make the example more practical

Reviewer: Or could rename it to 'big data' or smth

    process {

>       executor = 'slurm'

Reviewer: ```suggestion
```

Won't work on a laptop

>       queue    = 'standard'

Reviewer: ```suggestion
```

      cpus     = 4

>       memory   = 16.GB

Reviewer: ```suggestion
      memory   = 8.GB
```

To fit on a smaller laptop

    }
  }
}
```

Activate it alongside the existing profiles:

```bash

> nextflow run nf-core/demo -profile test,docker,myserver --outdir results

Reviewer: ```suggestion
nextflow run nf-core/demo -profile test,docker,myserver --outdir results_customprofile
```

```

:::note

> The `myserver` profile is illustrative — it will only run successfully if you have access to a SLURM cluster with a `standard` queue.
> On a workstation without SLURM, Nextflow will fail when it tries to submit jobs.
> Drop the `executor` and `queue` lines (or switch `executor` to `'local'`) to try the profile locally.
> :::

Reviewer: ```suggestion
```

I think just playing with the CPUs/Mem is sufficient demonstration, and allows people to activate play wit hit. if you want to find something else to configure (maybe a temp directory?) we can explore other options.

### Use shared institutional configs

If you work on a shared HPC cluster or cloud platform, there's a good chance someone has already written a profile for it.
The [nf-core/configs](https://github.com/nf-core/configs) repository collects over 150 cluster-specific configs contributed by the community, covering the executor, queue, resource limits, container engine, scratch paths, module systems, and any other cluster-specific quirks needed to run nf-core pipelines well in that environment.

> Every nf-core pipeline loads these configs automatically — there's nothing to download or copy.

Reviewer: ```suggestion
Every nf-core pipeline loads these configs automatically — there's nothing to download or copy yourself.
```

Because technically they do get downloaded on runtime, just by Nextflow itself

> At run time, each pipeline fetches the [`nfcore_custom.config`](https://github.com/nf-core/configs/blob/master/nfcore_custom.config) from the `nf-core/configs` repository and makes every profile in it available alongside the pipeline's own profiles.

Reviewer: Maybe  unnecessary detail? But don't mind
```suggestion
```

Some institutions also contribute **pipeline-specific** overrides — a profile that further tunes resources for a particular pipeline on a particular cluster.
These live under [`conf/pipeline/<pipeline>/<institution>.config`](https://github.com/nf-core/configs/tree/master/conf/pipeline) in the repository and are picked up automatically when you run that pipeline with the matching `-profile`.

:::tip
If your cluster doesn't have a profile yet, the [nf-core/configs contribution guide](https://github.com/nf-core/configs#documentation) describes how to add one.
Contributing a profile to `nf-core/configs` is the recommended way to share cluster configuration with your team — it removes the need for everyone to carry a `custom.config` file around, and benefits every nf-core pipeline at once.
:::

:::note
The shared configs are loaded over the network at run time.
If you're working on an air-gapped system, see [Running pipelines offline](../run-pipelines-offline.md) for how to download them ahead of time and point Nextflow at the local copy via the `custom_config_base` parameter.
:::

### Pass a config explicitly with `-c`

> For configuration you don't want loaded by default. For example, a one-off resource bump, a shared institutional config, or an experimental override, pass it on the command line with `-c`.

Reviewer: ```suggestion
For configuration you don't want loaded by default.
For example, a one-off resource bump, a shared institutional config, or an experimental override, pass it on the command line with `-c`.
```

> 1. Create a small `custom.config`:

Reviewer: ```suggestion
1. Create a small file named `custom.config`:
```

I find this makes it a bit clearer that it's not some special format per se

   ```groovy title="custom.config"
   process {
     cpus = 2

>      memory = 4.GB

Reviewer: ```suggestion
     cpus = 3
     memory = 6.GB
```

To vary a bit again

   }
   ```

2. Apply it with `-c`:

   ```bash

>    nextflow run nf-core/demo -profile test,docker -c custom.config --outdir results

Reviewer: ```suggestion
   nextflow run nf-core/demo -profile test,docker -c custom.config --outdir results_customconfig
```

   ```

> 3. Open `results/pipeline_info/execution_report.html` and inspect the resource columns to confirm the new requests were applied.

Reviewer: ```suggestion
3. Open `results_customconfig/pipeline_info/execution_report_<timestamp>.html` and inspect the resource columns to confirm the new requests were applied.
```

Yeah I think again this should have been introduced earlier, but I'm still not seeing any changes 😅 

There isn't anything in [`modules.config`](https://github.com/nf-core/demo/blob/master/conf/modules.config) so I assume it's coming from the base.conf labels?

Maybe we could add to `nf-core/demo` a single local module without a label, and this can be used for  demonstrating changes we are applying with the profiles/configs?

You can pass `-c` more than once.
Files are applied in order, so later ones override earlier one.

### Tune how the pipeline executes

> Beyond resource requests, config files control **how** each run executes

Reviewer: ```suggestion
Beyond resource requests, config files control **how** each run executes.
```

For example, which scheduler picks up jobs, what to do when a process fails, and where intermediate files live.
These settings sit alongside the `process` scope you used above, with a few sibling top-level options.

Edit `custom.config` to switch to a SLURM executor with automatic retries and a scratch work directory:

```groovy title="custom.config"

> workDir = '/scratch/$USER/nf-work'

Reviewer: For a fully interactive tutorial, could be set to `/tmp/foo` or maybe just within the working directory? (`$(pwd)`) or something?

process {
  executor      = 'slurm'

>   queue         = 'long'

Reviewer: I would again drop this for a fully interactive tutorial

  errorStrategy = 'retry'
  maxRetries    = 2
}
```

Apply it the same way as before:

```bash
nextflow run nf-core/demo -profile test,docker -c custom.config --outdir results
```

:::note
This config will only run successfully on a SLURM cluster with a `long` queue and access to `/scratch`.

> On a workstation, leave `executor` unset (Nextflow defaults to `local`), drop `queue`, and point `workDir` at a path you can actually write to, or simply skip running this config and read the annotations below.

Reviewer: ```suggestion
You will see now that a `nf-work/` directory specified in the config was generated in the directory where you executed the command, alongside the results directory.
```

:::

> Each setting controls a different aspect of the run:

Reviewer: ```suggestion
Each setting controls a different aspect of the run.
Common settings are:
```

> - **`process.executor`**: where jobs go — `local` (default), `slurm`, `awsbatch`, `lsf`, `pbs`, `kubernetes`, and so on.

Reviewer: I sort of feel changing the syntax `process {}` to `process.` to refer to the options, might be a source of potential confusion, but I assume you have more experience in this (or more faith in readers than me that they will work it out themsemselves 😬).

Author: I think keep as is, but noted and we can review again.

- **`process.queue`**: which queue or partition to submit to on shared clusters.
- **`process.errorStrategy`** with **`process.maxRetries`**: retry transient failures (for example, a node going down or a timeout) instead of failing the whole run.
- **`workDir`**: where Nextflow stages intermediate files — point it at fast scratch storage on HPC to keep your home filesystem clean.

:::tip

> For the full list of config scopes (`process`, `executor`, `docker`, `singularity`, `aws`, `azure`, `google`, `report`, `trace`, `timeline`, and more), see the [Nextflow config reference](https://docs.seqera.io/nextflow/reference/config).

Reviewer: ```suggestion
For the full list of config scopes and options (`process`, `executor`, `docker`, `singularity`, `aws`, `azure`, `google`, `report`, `trace`, `timeline`, and more), see the [Nextflow config reference](https://docs.seqera.io/nextflow/reference/config).
```

nf-core pipelines already enable the standard execution reports under `pipeline_info/`, so you only need to override those if you want custom paths.
:::

### Target specific processes with `withName` and `withLabel`

> The blanket `process { cpus = 2 }` example above applies to every process in the pipeline.

Reviewer: ```suggestion
The blanket `process { cpus = 2 }` examples above applies to every process in the pipeline.
```

> In practice you'll usually want to target specific steps.

Reviewer: ```suggestion
In practice you'll usually want to target specific steps of the pipeline.
```

> For example, bump the resources for an aligner without changing anything else, or pass an extra command-line argument to one tool.

Reviewer: ```suggestion
For example, to bump the resources for an aligner without changing anything else, or pass an extra command-line argument to one tool.
```

Nextflow gives you two selectors for this:

- **`withName`** matches a specific process by its fully qualified name. `nf-core/demo` runs three processes — [`FASTQC`](https://github.com/nf-core/demo/blob/master/modules/nf-core/fastqc/main.nf), [`SEQTK_TRIM`](https://github.com/nf-core/demo/blob/master/modules/nf-core/seqtk/trim/main.nf), and [`MULTIQC`](https://github.com/nf-core/demo/blob/master/modules/nf-core/multiqc/main.nf) — and each has a name you can see in `.nextflow.log` or in `pipeline_info/execution_trace.txt` after a run.

> - **`withLabel`** matches every process that carries a given label. All nf-core pipelines tag their processes with size labels (`process_low`, `process_medium`, `process_high`, `process_high_memory`) — see how they map to resource requests in [`nf-core/demo`'s `conf/base.config`](https://github.com/nf-core/demo/blob/master/conf/base.config) — so a single `withLabel` block can adjust whole resource tiers at once.

Reviewer: ```suggestion
- **`withLabel`** matches every process that carries a given label. All nf-core pipelines tag their processes with size labels (`process_low`, `process_medium`, `process_high`, `process_high_memory`) — see how they map to resource requests in [`nf-core/demo`'s `conf/base.config`](https://github.com/nf-core/demo/blob/master/conf/base.config). A single `withLabel` block can adjust whole resource tiers at once.
```

I don't know what you mean by 'whole resource tiers' here... maybe 'whole groups of similar modules at once' or something?

Edit your `custom.config` to use both:

```groovy title="custom.config"
process {
  withName: 'NFCORE_DEMO:DEMO:FASTQC' {
    cpus = 4
    memory = 8.GB
  }

  withLabel: 'process_low' {
    cpus = 2
    memory = 4.GB
  }
}
```

> Re-run with `-c custom.config` and check `results/pipeline_info/execution_report.html`.

Reviewer: ```suggestion
Re-run with `-c custom.config` and check `results_customlabels/pipeline_info/execution_report.html`.
```

But maybe include the full command if you agree

Reviewer: But still not working, WTF is going on ...

<img width="1166" height="862" alt="Image" src="https://github.com/user-attachments/assets/2cf310a1-c454-4823-aaf1-29b9c114e173" />

Am I being stupid?

```
nextflow run nf-core/demo -profile test,docker -c custom.config --outdir results_customlabels
```

with the `custom.config`

```

   process {
     cpus = 3 
     memory = 6.GB
   }

workDir = "/${System.getenv('PWD')}/nf-work"
process {
  errorStrategy = 'retry'
  maxRetries    = 2
}

process {
  withName: 'NFCORE_DEMO:DEMO:FASTQC' {
    cpus = 4
    memory = 8.GB
  }

  withLabel: 'process_low' {
    cpus = 2
    memory = 4.GB
  }
}

```

FASTQC should report the resources you set, and every other low-tier process should pick up the `process_low` block.

:::tip

> Find the full process name for `withName` in `.nextflow.log` or in `pipeline_info/execution_trace.txt` after a run.

Reviewer: ```suggestion
Find the full process name for `withName` in the console during a run, or `.nextflow.log` or in `pipeline_info/execution_trace.txt` after a run.
```

Process labels are declared on each `process` definition inside the module.
For example, [`FASTQC`'s `main.nf`](https://github.com/nf-core/demo/blob/master/modules/nf-core/fastqc/main.nf) declares `label 'process_medium'` on its second line.
The size labels (`process_low`, `process_medium`, `process_high`, `process_high_memory`) are standard across all nf-core pipelines.
:::

### Pass tool arguments with `ext.args`

`ext.args` is a per-process configuration value that nf-core modules use to inject extra command-line flags into the underlying tool.
It lives inside a `withName` selector (so it follows the same targeting rules as the previous section), and the module picks it up at runtime with `def args = task.ext.args ?: ''`.

> By convention, nf-core pipelines collect every per-process option in a dedicated file, [`conf/modules.config`](https://github.com/nf-core/demo/blob/master/conf/modules.config). It is another source of configuration loaded automatically through the pipeline's `nextflow.config`.

Reviewer: ```suggestion
By convention, nf-core pipelines collect every per-process option in a dedicated file, [`conf/modules.config`](https://github.com/nf-core/demo/blob/master/conf/modules.config).
It is another source of configuration loaded automatically through the pipeline's `nextflow.config`.
```

> Open `nf-core/demo`'s copy to see two real uses of `ext.args`:

Reviewer: ```suggestion
Inspect [`nf-core/demo`'s copy](https://github.com/nf-core/demo/blob/45904cb9d12db3d89900e6c479fe604ef71b297b/conf/modules.config#L21-L28) to see two real uses of `ext.args`:
```

```groovy
withName: FASTQC {
    ext.args = '--quiet'
    // ...
}

withName: 'MULTIQC' {
    ext.args = { params.multiqc_title ? "--title \"$params.multiqc_title\"" : '' }
    // ...
}
```

> FASTQC always runs with `--quiet`.

Reviewer: ```suggestion
FASTQC always runs with FastQC's `--quiet` mode.
```

MULTIQC dynamically picks up `--title` if the `multiqc_title` parameter is set.
This is why the `multiqc_title` value from your `params.yaml` flowed through into the MultiQC report.

To pass your own flags without forking the module, override `ext.args` in `custom.config`:

```groovy title="custom.config"
process {
  withName: 'NFCORE_DEMO:DEMO:FASTQC' {
    ext.args = '--quiet --noextract'
  }
}
```

Re-run with `-c custom.config` and FASTQC will pick up the new flags through `task.ext.args` inside its [`main.nf`](https://github.com/nf-core/demo/blob/master/modules/nf-core/fastqc/main.nf).

> Sibling keys you'll see in the same file:

Reviewer: ```suggestion
Re-run with `-c custom.config` and FASTQC will pick up the new flags through `task.ext.args` inside its [`main.nf`](https://github.com/nf-core/demo/blob/master/modules/nf-core/fastqc/main.nf).

:::danger
Customising `ext.args` is not generally not recommended, and may break the pipeline as they have not been tested by the developer.
It is recommended to request official support for a new tool option or argument from the pipeline developer.
Customise `ext.args` in a config as a last resort. 
:::

Sibling keys you'll see in the same file:
```

- **`ext.args2`, `ext.args3`** — second and third argument sets for modules that call more than one tool.
- **`ext.prefix`** — overrides the prefix used for the module's output filenames.

:::tip
`ext.args` is the right way to pass tool-specific flags that aren't exposed as pipeline parameters.

> Never edit the module's `main.nf` to add a flag — `ext.args` exists precisely so you can extend the command line without diverging from the canonical pipeline.

Reviewer: ```suggestion
Never edit the module's `main.nf` to add a flag — `ext.args` exists precisely so you can extend the command line without changing the versioned pipeline code.
```

:::

For `resourceLimits`, executor tuning, and container registry overrides, see [System requirements](../configuration/nextflow-for-your-system.md).
For profile precedence and shared institutional configs, see [Configuration options](../configuration/configuration-options.md).

## Put it all together

Each layer you've explored solves a different problem.
In practice, a single run usually uses several at once.
This step combines them into one invocation so you can see how they interact.

> 

Reviewer: Missing command line parameters?

> 1. Set environment variables for runtime concerns that don't change between runs:

Reviewer: I think including this at the top of the order makes now, as you've introduced when to use them above, so a user should understand the flow of 'set once and never look again, share across pipelines, across runs, one off run' etc.

   ```bash
   export NXF_VER=24.10.5
   export NXF_WORK=$HOME/nf-demo-work
   ```

2. Capture pipeline parameters in a `params.yaml` file so the run is reproducible:

   ```yaml title="params.yaml"
   outdir: my_results
   multiqc_title: "nf-core/demo configured run"
   ```

3. Put per-process overrides in `custom.config`:

   ```groovy title="custom.config"
   process {
     withName: 'NFCORE_DEMO:DEMO:FASTQC' {
       cpus = 4
       memory = 8.GB
       ext.args = '--quiet --noextract'
     }

     withLabel: 'process_low' {
       cpus = 2
       memory = 4.GB
     }
   }
   ```

4. Launch the run, activating the `test` and `docker` profiles and passing the parameter and config files:

   ```bash
   nextflow run nf-core/demo \
     -profile test,docker \
     -params-file params.yaml \
     -c custom.config
   ```

> This single command exercises every layer the tutorial introduced. Each one is resolved independently:

Reviewer: I didn't check the command because of my comments above, but the concept sounds good
```suggestion
This single command exercises every layer the tutorial introduced.
Each one is resolved independently:
```

- **Runtime**: `NXF_VER` and `NXF_WORK` are read from the shell before Nextflow parses any config, pinning the version and redirecting intermediate files.
- **Configuration**: config files layer in a fixed order — the pipeline's bundled `nextflow.config` (which pulls in `conf/base.config` and `conf/modules.config`) loads first, the `test` and `docker` profiles override matching keys, and `custom.config` overrides them last. The `withName` and `withLabel` blocks in `custom.config` therefore win for FASTQC and every `process_low` step.
- **Parameters**: values in `params.yaml` set the pipeline parameters. A matching `--flag` on the command line would override them.

To confirm each layer took effect, check:

> - `my_results/pipeline_info/execution_report.html` — FASTQC should report 4 CPUs and 8 GB of memory.

Reviewer: Hasn't been fore me...

- `my_results/multiqc/multiqc_report.html` — the report title should read "nf-core/demo configured run".
- `$HOME/nf-demo-work/` — intermediate files should appear here instead of under `./work/`.

:::tip
A rule of thumb for where each setting belongs:

- **Environment variables** for runtime concerns tied to your machine or session (Nextflow version, cache locations, offline mode).
- **`params.yaml`** for any setting the pipeline exposes as a parameter.
- **Profiles** for reusable bundles you switch on by name (container engine, test data, institutional cluster).
- **`custom.config`** for one-off or team-specific overrides that don't warrant a profile.

>   :::

Reviewer: ```suggestion
:::
```

- Activated profiles to switch between container engines and test scenarios.
- Layered a custom `custom.config` file with process-specific overrides.
- Used `ext.args` (the convention for `conf/modules.config`) to extend a tool's command line without editing the module.
- Used `NXF_*` environment variables to control Nextflow's runtime.
- Understood which configuration source overrides which.

:::note{title="Prerequisites"}
You will need the following to get started:

- An internet connection
- [Nextflow version 25.10.4 or later](../../get_started/environment_setup/nextflow.md)
- A container engine, such as Docker, Singularity, or Conda. See [Software dependencies](../../get_started/environment_setup/software-dependencies.md)

:::

## Baseline run

Before changing anything, run the pipeline with no custom configuration so you have something to compare against.
This baseline run relies entirely on `nf-core/demo`'s built-in defaults.
No environment variables, no custom parameters, and no extra config files are set.
In each of the following steps you will replace one of these defaults and observe the effect.

> Start with a plain test invocation as your reference point:

Reviewer: ```suggestion
Start with a plain test invocation as your reference point, replacing `docker` with your preferred software dependency manager (for example, `singularity` or `conda`).
```

```bash
nextflow run nf-core/demo -profile test,docker --outdir results
```

> Replace `docker` with your preferred software dependency manager (for example, `singularity` or `conda`).

Reviewer: ```suggestion
```

For example, inputs, outputs, skip flags, reference data choices, and tool toggles.
They're documented per pipeline.
For `nf-core/demo`, see the [parameters reference](https://nf-co.re/demo/parameters) and [`nextflow_schema.json`](https://github.com/nf-core/demo/blob/master/nextflow_schema.json).

You can supply them two ways:

- Directly on the command line
- From a `params.yaml` or `params.json` file.

The two forms are interchangeable, and you can mix them.

:::important
CLI flags override values in a parameter file.
:::

1. Override a single parameter on the command line:

   ```bash
   nextflow run nf-core/demo -profile test,docker --outdir results_customcommandline
   ```

   Output now goes to `results_customcommandline/` instead of `results/`.

2. For longer parameter sets, create a `params.yaml` file:

   ```yaml title="params.yaml"
   outdir: results_customparamsfile
   multiqc_title: "nf-core/demo parameter file configured run"
   ```

3. Apply it with `-params-file`:

   ```bash
   nextflow run nf-core/demo -profile test,docker -params-file params.yaml
   ```

>    Open `results_customparamsfile/multiqc/multiqc_report.html` to confirm the report title was set.

Reviewer: ```suggestion
   Open `results_customparamsfile/multiqc/nf-coredemo-parameter-file-configured-run_multiqc_report.html` to confirm the report title was set.
```

### Auto-loaded `nextflow.config` files

Nextflow looks for config files in three places without being told.

1. **The pipeline's own `nextflow.config` in the project directory.** Included with every nf-core pipeline's source code. For example, [`nf-core/demo`'s `nextflow.config`](https://github.com/nf-core/demo/blob/master/nextflow.config), which loads [`conf/base.config`](https://github.com/nf-core/demo/blob/master/conf/base.config) and declares every shipped profile. **Don't edit this file**. See the warning in [Configuration options](../configuration/configuration-options.md).
2. **`$HOME/.nextflow/config`.** Your personal config, applied to every Nextflow run you launch. Good for settings that should follow you across all pipelines, such as your Singularity cache location or a default executor for your workstation.
3. **`nextflow.config` in the launch directory.** Applied automatically whenever you run Nextflow from that directory. Good for per-project or per-working-directory settings you don't want to retype each time.

Add a `nextflow.config` in your current directory:

```groovy title="nextflow.config"
process {
  cpus = 1
  memory = 2.GB
}
```

> Re-run the baseline command. No `-c` flag is needed, Nextflow picks the file up automatically:

Reviewer: A bit annoying we don't have something to confirm this still...

Author: I still find this weird. Need to double-check.

Reviewer: I worked out (see previous comments about labels overwrriding this. SHould be fixed with the new version of demo)

```bash
nextflow run nf-core/demo -profile test,docker --outdir results_customnextflowconfig
```

:::tip
Use `$HOME/.nextflow/config` for personal defaults that follow you across machines, and a launch-directory `nextflow.config` for project-specific settings.

> Reserve `-c` (covered below) for one-off overrides and shared team configs that should live alongside your pipeline command.

Reviewer: ```suggestion
Reserve `-c` (covered below) for one-off overrides and shared project configs that should live alongside your pipeline command.
```

Not sure what you mean by 'team configs', does 'projects' work still?

Every nf-core pipeline comes with a standard set of profiles:

- **Software profiles**: `docker`, `singularity`, `apptainer`, `podman`, `conda`, `charliecloud`, `shifter` (one per container engine or environment manager).
- **Test profiles**: [`test`](https://github.com/nf-core/demo/blob/master/conf/test.config) (a small dataset for quick verification) and [`test_full`](https://github.com/nf-core/demo/blob/master/conf/test_full.config) (the full-size dataset used in CI).

  :::note
  The [`test_full`](https://github.com/nf-core/demo/blob/master/conf/test_full.config) profile in nf-core/demo is also a very small test dataset that can be used for testing.
  In other pipelines these can be much larger, but produce realistic output.
  :::

- **Institutional profiles**: contributed to [nf-core/configs](https://github.com/nf-core/configs) and loaded automatically by every nf-core pipeline. Activate one with `-profile <institution>` if your cluster has one - see [Use shared institutional configs](#use-shared-institutional-configs).

Combine profiles with commas, and remember that order matters — later profiles override earlier ones where they overlap:

```bash

> nextflow run nf-core/demo -profile test,docker --outdir results

Reviewer: Are users meant to run this? If not we should have a warning

On a workstation without those resources available, Nextflow will fail when it tries to run the processes.
:::

### Use shared institutional configs

If you work on a shared HPC cluster or cloud platform, there's a good chance someone has already written a profile for it.
The [nf-core/configs](https://github.com/nf-core/configs) repository collects over 150 cluster-specific configs contributed by the community, covering the executor, queue, resource limits, container engine, scratch paths, module systems, and any other cluster-specific quirks needed to run nf-core pipelines well in that environment.

Every nf-core pipeline loads these configs automatically — there's nothing to download or copy yourself.
At run time, each pipeline fetches the [`nfcore_custom.config`](https://github.com/nf-core/configs/blob/master/nfcore_custom.config) from the `nf-core/configs` repository and makes every profile in it available alongside the pipeline's own profiles.
You activate one the same way you activate any other profile:

```bash
nextflow run nf-core/demo -profile test,docker,<institution> --outdir results
```

> Replace `<institution>` with the profile name for your environment.

Reviewer: ```suggestion

```

A handful of examples from the [configs directory](https://github.com/nf-core/configs/tree/master/conf):

- [`uppmax`](https://github.com/nf-core/configs/blob/master/conf/uppmax.config) — UPPMAX (Sweden)
- [`bih`](https://github.com/nf-core/configs/blob/master/conf/bih.config) — Berlin Institute of Health
- [`aws_tower`](https://github.com/nf-core/configs/blob/master/conf/aws_tower.config) — Seqera Platform on AWS
- [`crick`](https://github.com/nf-core/configs/blob/master/conf/crick.config) — Francis Crick Institute

> Browse the full list at [nf-co.re/configs](https://nf-co.re/configs).

Reviewer: ```suggestion
Browse the full list at [nf-co.re/configs](https://nf-co.re/configs).

You activate one the same way you activate any other profile. If your institution's infrastructure is supported by nf-core/configs, replace `<institution>` with the profile name for your environment to try it out.

```bash
nextflow run nf-core/demo -profile test,docker,<institution> --outdir results
```
```

Contributing a profile to `nf-core/configs` is the recommended way to share cluster configuration with your team — it removes the need for everyone to carry a `custom.config` file around, and benefits every nf-core pipeline at once.
:::

:::note
The shared configs are loaded over the network at run time.
If you're working on an air-gapped system, see [Running pipelines offline](../run-pipelines-offline.md) for how to download them ahead of time and point Nextflow at the local copy via the `custom_config_base` parameter.
:::

### Pass a config explicitly with `-c`

For configuration you don't want loaded by default.
For example, a one-off resource bump, a shared institutional config, or an experimental override, pass it on the command line with `-c`.

1. Create a small file named `custom.config`:

   ```groovy title="custom.config"
   process {
     cpus = 3
     memory = 6.GB
   }
   ```

2. Apply it with `-c`:

   ```bash
   nextflow run nf-core/demo -profile test,docker -c custom.config --outdir results_customconfig
   ```

> 3. Open `results_customconfig/pipeline_info/execution_report_<timestamp>.html` and inspect the resource columns to confirm the new requests were applied

Reviewer: Yeah this doesn't work until we have a unlabelled process (like in the original `nextflow.config` stuff above)

You can pass `-c` more than once.
Files are applied in order, so later ones override earlier ones.

### Tune how the pipeline executes

Beyond resource requests, config files control **how** each run executes.
For example, which scheduler picks up jobs, what to do when a process fails, and where intermediate files live.
These settings sit alongside the `process` scope you used above, with a few sibling top-level options.

Edit `custom.config` to add automatic retries and redirect the work directory:

```groovy title="custom.config"
workDir = 'nf-work'

process {

>   errorStrategy = 'retry'

Reviewer: ```suggestion

  cpus = 3
  memory = 6.GB
  errorStrategy = 'retry'
```

To ensure it includes the original stuff

  maxRetries    = 2
}
```

Apply it the same way as before:

```bash

> nextflow run nf-core/demo -profile test,docker -c custom.config --outdir results

Reviewer: ```suggestion
nextflow run nf-core/demo -profile test,docker -c custom.config --outdir results_customconfig
```

```

:::note
You will now see that an `nf-work/` directory specified in the config was generated in the directory where you executed the command, alongside the results directory.
:::

Each setting controls a different aspect of the run.
Common settings are:

- **`process.executor`**: where jobs go — `local` (default), `slurm`, `awsbatch`, `lsf`, `pbs`, `kubernetes`, and so on.
- **`process.queue`**: which queue or partition to submit to on shared clusters.
- **`process.errorStrategy`** with **`process.maxRetries`**: retry transient failures (for example, a node going down or a timeout) instead of failing the whole run.
- **`workDir`**: where Nextflow stages intermediate files — point it at fast scratch storage on HPC to keep your home filesystem clean.

> :::tip

Reviewer: ````suggestion
- **`workDir`**: where Nextflow stages intermediate files — point it at fast scratch storage on HPC to keep your home filesystem clean.

:::info
The syntax

```groovy
process.queue = small
```

is equivalent to

```groovy
 process {
    queue = 'small'
}
``` 
:::

:::tip
````

- **`withLabel`** matches every process that carries a given label. All nf-core pipelines tag their processes with size labels (`process_low`, `process_medium`, `process_high`, `process_high_memory`) — see how they map to resource requests in [`nf-core/demo`'s `conf/base.config`](https://github.com/nf-core/demo/blob/master/conf/base.config). A single `withLabel` block can adjust whole resource tiers at once.

Edit your `custom.config` to use both:

```groovy title="custom.config"
process {
  withName: 'NFCORE_DEMO:DEMO:FASTQC' {
    cpus = 4
    memory = 8.GB
  }

  withLabel: 'process_low' {
    cpus = 2
    memory = 2.GB
  }
}
```

Re-run with `-c custom.config` and check `pipeline_info/execution_report.html` in your output directory.
FASTQC should report the resources you set, and every other low-tier process should pick up the `process_low` block.

:::tip

> Find the full process name for `withName` in the console during a run, or `.nextflow.log` or in `pipeline_info/execution_trace.txt` after a run.

Reviewer: ````suggestion

```bash
nextflow run nf-core/demo -profile test,docker -c custom.config --outdir results_customconfig
```

:::tip
Find the full process name for `withName` in the console during a run, or `.nextflow.log` or in `pipeline_info/execution_trace.txt` after a run.
````

It is another source of configuration loaded automatically through the pipeline's `nextflow.config`.
Open `nf-core/demo`'s [`conf/modules.config`](https://github.com/nf-core/demo/blob/45904cb9d12db3d89900e6c479fe604ef71b297b/conf/modules.config#L21-L28) to see two real uses of `ext.args`:

```groovy
withName: FASTQC {
    ext.args = '--quiet'
    // ...
}

withName: 'MULTIQC' {
    ext.args = { params.multiqc_title ? "--title \"$params.multiqc_title\"" : '' }
    // ...
}
```

FASTQC always runs with FastQC's `--quiet` mode.
MULTIQC dynamically picks up `--title` if the `multiqc_title` parameter is set.
This is why the `multiqc_title` value from your `params.yaml` flowed through into the MultiQC report.

To pass your own flags without forking the module, override `ext.args` in `custom.config`:

```groovy title="custom.config"
process {
  withName: 'NFCORE_DEMO:DEMO:FASTQC' {
    ext.args = '--quiet --noextract'
  }
}
```

Re-run with `-c custom.config` and FASTQC will pick up the new flags through `task.ext.args` inside its [`main.nf`](https://github.com/nf-core/demo/blob/master/modules/nf-core/fastqc/main.nf).

> 

Reviewer: ```suggestion

```bash
nextflow run nf-core/demo -profile test,docker -c custom.config --outdir results_customconfig
```

```

   ```

   The version line at the top of stdout (and in `.nextflow.log`) should now report `25.10.4`, and intermediate files should appear under `$HOME/nf-demo-work/` instead of `./work/`.

2. To set a variable for a single command without exporting it, prefix the invocation:

   ```bash
   NXF_VER=25.10.4 nextflow run nf-core/demo -profile test,docker --outdir results
   ```

3. To persist a variable across sessions, add the export to your shell config:

   ```bash title=".bashrc"
   export NXF_VER=25.10.4
   ```

:::tip
The `NXF_*` variables nf-core users reach for most often are:

- `NXF_VER`: Pin a specific Nextflow version
- `NXF_HOME`: Override Nextflow's home directory (defaults to `$HOME/.nextflow`)
- `NXF_WORK`: Override the work directory for intermediate files
- `NXF_OFFLINE`: Run without contacting the network (see [Running pipelines offline](../run-pipelines-offline.md))
- `NXF_SINGULARITY_CACHEDIR` / `NXF_APPTAINER_CACHEDIR`: Reuse downloaded container images across runs

For the full list, see the [Nextflow environment variables reference](https://docs.seqera.io/nextflow/reference/env-vars).
:::

:::note

> `NXF_*` variables are distinct from pipeline parameters. They configure how Nextflow runs, not what the pipeline does. You cannot set `--input` or `--outdir` with an environment variable.

Reviewer: ```suggestion
`NXF_*` variables are distinct from pipeline parameters.
They configure how Nextflow runs, not what the pipeline does.
You cannot set `--input` or `--outdir` with an environment variable.
```

:::

## Put it all together

Each layer you've explored solves a different problem.
In practice, a single run usually uses several at once.
This step combines them into one invocation so you can see how they interact.

1. Set environment variables for runtime concerns that don't change between runs:

   ```bash
   export NXF_VER=25.10.4

>    export NXF_WORK=$HOME/nf-demo-work

Reviewer: I personally wouldn't want to pollute someones home directory with a tutorial it should remain selfcontained... should we start saying 'make a directory called `nf-core-configuration-tutorial` and have evertyhign running from within there? Then they can easily delete at the end.

Author: Maybe we should set the ENV variable in the first step 😉 

But yes, we can say make a folder and do everything there.
